src/spacebro-client.js

```diff
@@ -17,171 +17,286 @@ let config = {
+        this.logger.warn(err)
+        this.events['reconnect_error'] && this.events['reconnect_error'].dispatch(err)
+      })
+      .on('reconnect_failed', (err) => {
+        this.logger.warn('socket cannot reconnect')
+        this.logger.warn(err)
+        this.events['reconnect_failed'] && this.events['reconnect_failed'].dispatch(err)
+      })
+
+      .on('*', ({ data }) => {
+        let [eventName, args] = data
+        if (!this.config.sendBack && args._from === this.config.clientName) {
+          return
+        }
+        if (this.events[eventName]) {
+          this.logger.log(`socket received ${eventName} with data:`, args)
+          for (let unpack of filterHooks(eventName, this.unpackers)) {
+            const unpacked = unpack({ eventName, data: args })
+            args = unpacked || args
+          }
+          this.events[eventName].dispatch(args)
         }
-        events[eventName].dispatch(args)
+      })
+  }
+
+  disconnect () {
+    if (this.socket) {
+      this.socket.disconnect()
+    }
+    this.connected = false
+    this.unpackers = []
+    this.packers = []
+  }
+
+  addPacker (handler, priority, eventName) {
+    this.packers.push({ eventName, handler, priority })
+  }
+  addUnpacker (handler, priority, eventName) {
+    this.unpackers.push({ eventName, handler, priority })
+  }
+
+  emit (eventName, data = {}) {
+    // null is a type of Object. so we have to check null and undefined with loosy compare
+    if (typeof data !== 'object' || data == null) {
+      data = {data: data}
+      data.altered = true
+    }
+    this.sendTo(eventName, null, data)
+  }
+
+  sendTo (eventName, to = null, data = {}) {
+    if (this.connected) {
+      data._to = to
+      data._from = this.config.clientName
+      for (let pack of filterHooks(eventName, this.packers)) {
+        data = pack({eventName, data}) || data
       }
-    })
+      this.socket.emit(eventName, data)
+    } else {
+      this.logger.warn('can\'t emit, not connected.')
+    }
+  }
+
+  // Reception
+  on (eventName, handler, handlerContext, priority) {
+    if (this.events[eventName]) {
+      this.logger.warn(`Signal ${eventName} already exists`)
+    }
+    this.events[eventName] = new Signal()
+    this.events[eventName].add(handler, handlerContext, priority)
+  }
+
+  once (eventName, handler, handlerContext, priority) {
+    if (this.events[eventName]) {
+      this.logger.warn(`Signal ${eventName} already exists`)
+    }
+    this.events[eventName] = new Signal()
+    this.events[eventName].addOnce(handler, handlerContext, priority)
+  }
+
+  off (eventName) {
+    this.events[eventName].dispose()
+    delete this.events[eventName]
+  }
 }
 
-function addPacker (handler, priority, eventName) { addHook(packers, eventName, handler, priority) }
-function addUnpacker (handler, priority, eventName) { addHook(unpackers, eventName, handler, priority) }
+let lastSocket = null
+let fakeSocket = new SpacebroClient()
```

Comment:
It's a hack to store events added with the legacy `on` / `once` before `connect` is called.

```
const spacebroClient = require("spacebroClient")

spacebroClient.on("foobar", () => doThings())
spacebroClient.connect(...)
```